src/backend/distributed/utils/citus_nodefuncs.c

```diff
@@ -141,13 +145,45 @@ SetRangeTblExtraData(RangeTblEntry *rte, CitusRTEKind rteKind,
 }
 
 
+/*
+ * UnsetRangeTblExtraData changes an RTE that contains extradata back into a
+ * normal RTE_RELATION for the shard.
+ * IMPORTANT: This only works when the shard is present on the machine in
+ * question.
+ */
+void
+UnsetRangeTblExtraData(RangeTblEntry *rte)
```

Comment:
Can't we skip calling `UpdateRelationToShardNames()` during the planning so that we don't need this function and the traversal?

These two traversals show up in the perf report, so it'd be great to skip them.

That'd also significantly reduce the patch size. One idea could be move `UpdateRelationToShardNames` to `TaskQueryString ()`, and not call `UpdateRelationToShardNames()` during the planning.

Reply:
For fast path queries that we now can be locally executed we can do this. I think this would be good as a next PR on top of this one though.

Reply:
A form of this is now implemented in #3370